File: webhook_server.py
#!/usr/bin/env python3
"""
Webhook server for Telegram bot using Flask.
Receives webhook updates from Telegram and processes them.
"""

import logging
from flask import Flask, request, Response
from telegram import Update
from telegram.ext import Application
import asyncio
import json

from config import config
from bot import start_command, help_command, echo_message, error_handler

# Configure logging
logger = logging.getLogger(__name__)

# Create Flask app
app = Flask(__name__)

# Global bot application instance
bot_app: Application = None


async def process_update(update_data: dict) -> None:
    """Process a Telegram update asynchronously."""
    try:
        update = Update.de_json(update_data, bot_app.bot)
        await bot_app.process_update(update)
    except Exception as e:
        logger.error(f"Error processing update: {e}", exc_info=True)


@app.route('/')
def index():
    """Health check endpoint."""
    return {'status': 'ok', 'message': 'Telegram bot webhook server is running'}


@app.route(f'/webhook/<token>', methods=['POST'])
def webhook(token):
    """Handle incoming webhook updates from Telegram."""
    # Verify the token matches our bot token for security
    if token != config.bot_token:
        logger.warning(f"Webhook called with invalid token: {token[:10]}...")
        return Response(status=403)

    try:
        # Get the update data from Telegram
        update_data = request.get_json(force=True)
        logger.debug(f"Received webhook update: {json.dumps(update_data, indent=2)}")

        # Process the update asynchronously
        asyncio.run(process_update(update_data))

        return Response(status=200)

    except Exception as e:
        logger.error(f"Error in webhook endpoint: {e}", exc_info=True)
        return Response(status=500)


def setup_bot_application() -> Application:
    """Set up and configure the bot application."""
    logger.info("Setting up bot application for webhook mode...")

    # Create the Application
    application = Application.builder().token(config.bot_token).build()

    # Register command handlers
    from telegram.ext import CommandHandler, MessageHandler, filters
    application.add_handler(CommandHandler("start", start_command))
    application.add_handler(CommandHandler("help", help_command))

    # Register message handler for text messages
    text_filter = filters.TEXT & ~filters.COMMAND
    application.add_handler(MessageHandler(filters.UpdateType.MESSAGE & text_filter, echo_message))
    application.add_handler(MessageHandler(filters.UpdateType.EDITED_MESSAGE & text_filter, echo_message))

    # Register error handler
    application.add_error_handler(error_handler)

    logger.info("Bot handlers registered successfully")
    return application


async def set_webhook():
    """Set the webhook URL with Telegram."""
    webhook_url = f"{config.webhook_url}{config.webhook_path}/{config.bot_token}"

    logger.info(f"Setting webhook URL: {webhook_url}")

    try:
        await bot_app.bot.set_webhook(
            url=webhook_url,
            allowed_updates=["message", "edited_message"]
        )

        # Verify webhook was set
        webhook_info = await bot_app.bot.get_webhook_info()
        logger.info(f"Webhook set successfully!")
        logger.info(f"Current webhook URL: {webhook_info.url}")
        logger.info(f"Pending updates: {webhook_info.pending_update_count}")

    except Exception as e:
        logger.error(f"Failed to set webhook: {e}")
        raise


def main():
    """Main function to start the webhook server."""
    global bot_app

    logger.info("Starting Telegram bot in webhook mode...")

    try:
        # Set up bot application
        bot_app = setup_bot_application()

        # Initialize the bot application
        asyncio.run(bot_app.initialize())

        # Set webhook with Telegram
        asyncio.run(set_webhook())

        # Start Flask server
        logger.info(f"Starting Flask server on port {config.webhook_port}...")
        app.run(
            host='0.0.0.0',
            port=config.webhook_port,
            debug=False  # Set to False in production
        )

    except Exception as e:
        logger.critical(f"Failed to start webhook server: {e}")
        raise


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        logger.info("Webhook server stopped by user")
    except Exception as e:
        logger.critical(f"Critical error: {e}")
        exit(1)

File: js/webhookServer.js
const express = require('express');
const config = require('./config');
const { TokenStorage, handleOAuthCallback } = require('./googleAuth');
const { bot, setTokenStorage, handleDeletedMessage } = require('./bot');

const app = express();
app.use(express.json());

let tokenStorage;

app.get('/', (req, res) => {
  res.json({ status: 'ok', message: 'Telegram bot webhook server is running' });
});

app.get('/oauth/callback', async (req, res) => {
  const { code, state, error } = req.query;

  if (error) {
    console.warn(`OAuth error: ${error}`);
    return res.status(400).send(
      `<h1>Authentication Failed</h1><p>Error: ${error}</p><p>Please try /authenticate again in Telegram.</p>`
    );
  }

  if (!code || !state) {
    return res.status(400).send('<h1>Invalid Request</h1><p>Missing code or state parameter.</p>');
  }

  const userId = await handleOAuthCallback(
    code, state,
    config.googleClientId,
    config.googleClientSecret,
    config.googleRedirectUri,
    tokenStorage
  );

  if (!userId) {
    return res.status(400).send(
      '<h1>Authentication Failed</h1><p>Invalid or expired state. Please try /authenticate again in Telegram.</p>'
    );
  }

  try {
    await bot.sendMessage(userId,
      'Authentication successful! You can now send messages and they will be saved to your Google Drive.'
    );
  } catch (err) {
    console.error(`Failed to send confirmation to user ${userId}: ${err}`);
  }

  res.send('<h1>Authentication Successful!</h1><p>You can close this window and return to Telegram.</p>');
});

app.post(`${config.webhookPath}/${config.botToken}`, async (req, res) => {
  const update = req.body;
  try {
    if (update.deleted_messages) {
      const userIdFromChat = update.chat?.id;
      if (userIdFromChat) {
        for (const msg of update.deleted_messages) {
          if (msg.message_id) {
            await handleDeletedMessage(msg.message_id, userIdFromChat);
          }
        }
      }
    } else {
      bot.processUpdate(update);
    }
    res.sendStatus(200);
  } catch (err) {
    console.error(`Error processing update: ${err}`);
    res.sendStatus(500);
  }
});

async function setWebhook() {
  const webhookUrl = `${config.webhookUrl}${config.webhookPath}/${config.botToken}`;
  const redactedUrl = `${config.webhookUrl}${config.webhookPath}/[REDACTED]`;

  console.log(`Setting webhook URL: ${redactedUrl}`);
  try {
    await bot.setWebHook(webhookUrl, {
      allowed_updates: ['message', 'edited_message', 'message_delete'],
    });
    const info = await bot.getWebHookInfo();
    const loggedUrl = info.url ? info.url.replace(config.botToken, '[REDACTED]') : 'None';
    console.log(`Webhook set. URL: ${loggedUrl}, Pending updates: ${info.pending_update_count}`);
  } catch (err) {
    console.error(`Failed to set webhook: ${err}`);
    throw err;
  }
}

async function main() {
  console.log('Starting Telegram bot in webhook mode...');

  tokenStorage = new TokenStorage(config.databaseConfig, config.tokenEncryptionKey);
  setTokenStorage(tokenStorage);
  console.log('Token storage initialized');

  await setWebhook();

  app.listen(config.webhookPort, '0.0.0.0', () => {
    console.log(`Webhook server listening on port ${config.webhookPort}`);
  });
}

main().catch((err) => {
  console.error(`Failed to start: ${err}`);
  process.exit(1);
});
